### PR TITLE
First version of cognito

### PR DESCRIPTION
**Pull request by Dromadon**
This is a first version of having a cognito user pool + application clients.

It is not directly working as not linked to the API. But I would prefer to do this integration in the end2end test branch, what do you think @Mehdi-H ?

### CODE REVIEW

File: deploy/cognito/application_client.tf [hunk 1]
```diff
@@ -0,0 +1,48 @@
+// The E2E application client
```
Comment (Mehdi-H) [inline deploy/cognito/application_client.tf]: Do you want to keep this ? I think that the filename is pretty explicit

```suggestion
```

File: deploy/cognito/application_client.tf [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+  scope {
+    scope_name = "api"
+    scope_description = "full api access"
+  }
+
+  user_pool_id = aws_cognito_user_pool.users.id
+}
+
+// The webapp application client
```
Comment (Mehdi-H) [inline deploy/cognito/application_client.tf]: Maybe, putting this resource in a dedicated file would make its presence more explicit ?

File: deploy/cognito/application_client.tf (hunk 1 above)
Comment (Mehdi-H) [reply deploy/cognito/application_client.tf]: Also, I would name the file `client_application.tf`

File: deploy/cognito/userpool.tf [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    }
+  }
+
+  schema {
+    attribute_data_type = "String"
+    name = "phone"
+    required = false
+    mutable = false
+
+    string_attribute_constraints {
```
Comment (Mehdi-H) [inline deploy/cognito/userpool.tf]: Why did you put these values ? Just curious.

File: deploy/cognito/userpool.tf [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      max_length = 15
+      min_length = 3
```
Comment (Mehdi-H) [inline deploy/cognito/userpool.tf]: Also, I would feel it more natural to read with the minimum before the maximum, like
```suggestion
      min_length = 3
      max_length = 15
```

Comment (Mehdi-H) [reply deploy/cognito/userpool.tf]: Same comment for lines 19 & 20, what do you think ?

File: deploy/cognito/userpool.tf [hunk 5]
```diff
@@ -0,0 +1,38 @@
+variable env {}
+
+resource "aws_cognito_user_pool" "users" {
+  name = "hallebarde-${var.env}"
+
+  admin_create_user_config {
+    allow_admin_create_user_only = true
+  }
+
+  mfa_configuration = "OFF"
```
Comment (Mehdi-H) [inline deploy/cognito/userpool.tf]: Hm, is it legit Terraform HCL ^^ ? Where is it used ? I don't see any another occurence of it

File: deploy/cognito/application_client.tf (hunk 2 above)
Comment (Dromadon) [reply deploy/cognito/application_client.tf]: Yup, I will separate these two

File: deploy/cognito/userpool.tf (hunk 3 above)
Comment (Dromadon) [reply deploy/cognito/userpool.tf]: That is legacy from a previous test. I will delete the phone part (I am not 100% clear about keeping the email if we request the email oauth scope).

File: deploy/cognito/userpool.tf (hunk 4 above)
Comment (Dromadon) [reply deploy/cognito/userpool.tf]: OK

File: deploy/cognito/userpool.tf (hunk 5 above)
Comment (Dromadon) [reply deploy/cognito/userpool.tf]: Yeah it is, what bothers you?

Comment (Dromadon) [reply deploy/cognito/userpool.tf]: ping @Mehdi-H

Comment (Mehdi-H) [reply deploy/cognito/userpool.tf]: 1. I don't see where it is used
2. I generally see resource declaration in HCL, I don't think I ever saw simple key values declaration like that. It is not a problem, I just never saw this usage, just curious :d

Comment (Dromadon) [reply deploy/cognito/userpool.tf]: Mmmmh the mfa_configuration line is part of a resource declaration (the userpool)

**Pull request merged**